### PR TITLE
error in codegen if vendor provides kernel for a composite op with no…

### PR DESCRIPTION
**Pull request by bdhirsh**
Error if someone tries to modify the yaml by adding a composite op that would result in losing autograd for that op (because we can no longer use the composite formula).

This fails locally (correctly) on the staging branch because `dropout` is still in the yaml list, so I'll merge this after https://github.com/pytorch/pytorch/pull/66823 goes through.

Failure error:
```
AssertionError:
Found an entry for 'native_layer_norm' in the yaml. This operator is composite, which means that by default it will decompose into base operators. If decomposing is fine, then you can simply remove the entry from the yaml file. If you wish to provide an explicit kernel directly for this composite op, you can do so by:
 (a) moving the operator name under the 'autograd' entry in the yaml file
 (b) writing a custom Autograd Function for the operator
```

### CODE REVIEW

**Review by wconstab — APPROVED** [review]
awesome, thanks for doing this!  I was actually scratching my head last night thinking about this and it was a nice surprise that it was done for me when I woke up :)

**Comment by bdhirsh** [comment]
@wconstab I dropped the ball on this - merging it today, and hopefully we can merge it into master with the other LTC codegen changes in https://github.com/pytorch/pytorch/pull/69020

**Comment by bdhirsh** [comment]
Actually, I think I'll need to wait for https://github.com/pytorch/pytorch/pull/69169 to land before I can land this. `logical_not` is currently lowered by XLA and its marked (incorrectly) as `CompositeImplicitAutograd` in core, which is causing xla to trip on the codegen.

File: tools/codegen/gen_backend_stubs.py [hunk 1]
```diff
@@ -295,7 +316,7 @@ def make_file_manager(install_dir: str) -> FileManager:
         class_name = backend_indices[backend_dispatch_key].native_function_class_name()
 
         if impl_path is not None:
-            error_on_missing_kernels(native_functions, backend_indices, backend_key, autograd_key, impl_path)
+            error_on_missing_or_invalid_kernels(native_functions, backend_indices, backend_key, autograd_key, impl_path)
```
Comment (guoyejun) [inline tools/codegen/gen_backend_stubs.py:319]: for invalid kernel checking, it should be done no matter impl_path is provided or not.

we may move 'if impl_path is not None:' into function error_on_missing_or_invalid_kernels.

File: tools/codegen/gen_backend_stubs.py [hunk 2]
```diff
@@ -171,6 +171,27 @@ def create_decl(f: NativeFunction) -> str:
...
 
+    # Additionally, provide a nice error if a user tries to register a composite op with no derivative formula to a backend key.
+    # Otherwise, autograd will silently stop working for their op.
+    for func in native_functions:
+        if backend_indices[backend_key].has_kernel(func) and func.has_composite_implicit_autograd_kernel:
+            # At this point, we know that the vendor provided a kernel for an op that also has a CompositeImplicitAutograd kernel.
+            # What we really want to error on, though, is if that op doesn't have a derivative formula registered in-tree.
+            # We can guess that based on whether or not the op has ANY in-tree non-composite kernels.
+            # Technically this has a small chance of false negatives, but it's easier than re-parsing derivatives.yaml here.
+            idxs = [idx for idx in backend_indices.values() if not idx.external and idx.has_kernel(func)]
```
Comment (guoyejun) [inline tools/codegen/gen_backend_stubs.py:182]: idxs is not used nex

File: tools/codegen/gen_backend_stubs.py [hunk 3]
```diff
@@ -171,6 +171,23 @@ def create_decl(f: NativeFunction) -> str:
...
+            # We can guess that based on whether or not the op has ANY in-tree non-composite kernels.
+            # Technically this has a small chance of false negatives, but it's easier than re-parsing derivatives.yaml here.
+            has_non_composite_in_tree_backend = any(idx.has_kernel(f) for idx in backend_indices.values() if not idx.external)
+            assert not has_non_composite_in_tree_backend, \
+                f"\nFound an entry for '{f.func.name}' in the yaml. " \
+                "This operator is composite, which means that by default it will decompose into base operators. " \
+                "If decomposing is fine, then you can simply remove the entry from the yaml file. " \
+                "If you wish to provide an explicit kernel directly for this composite op, you can do so by:\n " \
+                "(a) moving the operator name under the 'autograd' entry in the yaml file\n " \
+                "(b) writing a custom Autograd Function for the operator"
```
[the remark this answers is not in the document]
Comment (guoyejun) [reply tools/codegen/gen_backend_stubs.py:193]: I have one open:  is it possible to first get a list of register-able OPs by scanning the backend_indices with in tree backends except DispatchKey.CompositeImplicitAutograd and DispatchKey.CompositeExplicitAutograd, and then check if the OPs registered by external backend are in the list or not.

The reason is that I want to get the list: the register-able OPs that we have not registered in our backend.  This would be easy if we have the list of all register-able OPs.

[the remark this answers is not in the document]
Comment (bdhirsh) [reply tools/codegen/gen_backend_stubs.py:193]: That depends on what you're looking for.

If you're interested in "what are all of the registerable ops", you can just look at `aten/src/ATen/native/native_functions.yaml`, which has the complete list.

The functions in the `DispatchKey.CompositeImplicitAutograd` and `DispatchKey.CompositeExplicitAutograd` probably aren't what you want, because those are ops that have composite implementations, a custom backend like XPU would get them for free (they would decompose into base ops, which you would then have to implement).

File: tools/codegen/gen_backend_stubs.py (hunk 1 above)
Comment (bdhirsh) [reply tools/codegen/gen_backend_stubs.py:319]: fair point. I'm going to re-make this PR and base it against master after #69020 lands, but I'll make that change when I do.

File: tools/codegen/gen_backend_stubs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (guoyejun) [reply tools/codegen/gen_backend_stubs.py:193]: yes, we can get it from native_functions.yaml, my thinking was to make this script provide a new feature (list un-registered OPs) by adding a little effort since we already has similar logic here.  It doesn't matter if we decide not to provide such OPs in this script, I'm open to write another one.

[the remark this answers is not in the document]
Comment (bdhirsh) [reply tools/codegen/gen_backend_stubs.py:193]: There's actually a more programmatic way you can figure out which ops have registrations.
```
import torch  # this runs all registrations to the dispatcher in core
import torch_ipex  # assuming this runs dispatcher registrations for XPU

# This prints EVERY operator that's registered to the dispatcher
torch._C._dispatch_print_registrations_for_dispatch_key()

# This prints every op with a CompositeImplicitAutograd registration
torch._C._dispatch_print_registrations_for_dispatch_key('CompositeImplicitAutograd')

# This prints every op with an XPU registration
torch._C._dispatch_print_registrations_for_dispatch_key('XPU')
```

File: tools/codegen/gen_backend_stubs.py [hunk 4]
```diff
@@ -171,6 +171,27 @@ def create_decl(f: NativeFunction) -> str:
...
+            # What we really want to error on, though, is if that op doesn't have a derivative formula registered in-tree.
+            # We can guess that based on whether or not the op has ANY in-tree non-composite kernels.
+            # Technically this has a small chance of false negatives, but it's easier than re-parsing derivatives.yaml here.
+            idxs = [idx for idx in backend_indices.values() if not idx.external and idx.has_kernel(func)]
+            has_non_composite_in_tree_backend = any(
+                idx.has_kernel(func) for idx in backend_indices.values()
+                if not idx.external and idx.dispatch_key != DispatchKey.CompositeImplicitAutograd
+                and idx.dispatch_key != DispatchKey.CompositeExplicitAutograd)
+            assert has_non_composite_in_tree_backend, \
+                f"\nFound an entry for '{func.func.name}' in the yaml. " \
```
Comment (guoyejun) [inline tools/codegen/gen_backend_stubs.py:188]: in the yaml for {backend_key}. " \

we may also print the dispatch key since one backend might have several dispatch keys.

File: tools/codegen/gen_backend_stubs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (guoyejun) [reply tools/codegen/gen_backend_stubs.py:193]: hi @bdhirsh , sorry for the late response, I tried "torch._C._dispatch_print_registrations_for_dispatch_key()" and found that all the ops in native_functions.yaml are returned, including ops DEPRECATED, ops with manual_cpp_binding=True, and other ops not registered with CUDA/CPU in native_functions.yaml.

```
# DEPRECATED. DO NOT USE
- func: _cast_Byte(Tensor self, bool non_blocking=False) -> Tensor
  variants: function

- func: _backward(Tensor self, Tensor[] inputs, Tensor? gradient=None, bool? retain_graph=None, bool create_graph=False) -> ()
  manual_cpp_binding: True
  variants: method

- func: _fw_primal(Tensor(a) self, int level) -> Tensor(a)
  variants: method
  dispatch:
    CompositeExplicitAutograd: _fw_primal

- func: _make_dual(Tensor(a) primal, Tensor tangent, int level) -> Tensor(a)
  variants: function

- func: refine_names(Tensor(a) self, Dimname[] names) -> Tensor(a)
  variants: method

- func: _sobol_engine_initialize_state_(Tensor(a!) self, int dimension) -> Tensor(a!)

- func: silu_backward(Tensor grad_output, Tensor self) -> Tensor
  structured_delegate: silu_backward.grad_input
  python_module: nn
  dispatch:
    CompositeImplicitAutograd: math_silu_backward

```

From an external backend's viewpoint, we want to get all the OPs we do need to register, any guidance for us to quickly filter the ops, thanks.
